File: sorting/part2/quicksort.py
''' this is quick sort algo
here this is also divide and conquer rule but more optimised tha merge sort

'''
def partition(arr,low,high):
    pivot = arr[low]
    i = low  
    j = high
    while i < j:
        while arr[i] <= pivot and i <= high:
            i+=1
        while arr[j] > pivot and j >= low +1:
            j-=1
        if i < j:
            arr[i],arr[j]=arr[j],arr[i]
    arr[low],arr[j] = arr[j],arr[low]
    return j
def quicksort(arr,low,high):
    if low < high:
        pindex = partition(arr,low,high)
        quicksort(arr,low,pindex-1)
        quicksort(arr,pindex+1,high)
arr = [4, 6, 2, 5, 7, 9, 1, 3]
print("Before using QuickSort:")
print(arr)

quicksort(arr,0,len(arr)-1)  # Sort the array in-place

print("After using QuickSort:")
print(arr)

File: sorting/part2/quicksort.cpp
#include <iostream>
using namespace std;

// Partition function for Quick Sort
int partition(int arr[], int low, int high) {
    int pivot = arr[low];
    int i = low, j = high;
    
    // Find the correct position for the pivot
    while (i < j) {
        // Move i to the right while arr[i] is less than or equal to pivot
        while (arr[i] <= pivot && i <= high - 1) {
            i++;
        }
        // Move j to the left while arr[j] is greater than pivot
        while (arr[j] > pivot && j >= low + 1) {
            j--;
        }
        // Swap arr[i] and arr[j] if they are not in the correct position
        if (i < j) {
            swap(arr[i], arr[j]);
        }
    }
    // Swap the pivot element with the element at index j
    swap(arr[low], arr[j]);
    return j;
}

// Quick Sort function
void quicksort(int arr[], int low, int high) {
    if (low < high) {
        int pindex = partition(arr, low, high); // Partition the array
        quicksort(arr, low, pindex - 1); // Recursively sort the left part
        quicksort(arr, pindex + 1, high); // Recursively sort the right part
    }
}

int main() {
    int arr[] = {10, 7, 8, 9, 1, 5};
    int n = sizeof(arr) / sizeof(arr[0]);

    cout << "Original array: ";
    for (int i = 0; i < n; i++) {
        cout << arr[i] << " ";
    }
    cout << endl;

    quicksort(arr, 0, n - 1); // Calling Quick Sort

    cout << "Sorted array: ";
    for (int i = 0; i < n; i++) {
        cout << arr[i] << " ";
    }
    cout << endl;

    return 0;
}
